nodewatcher/nodewatcher.py

```diff
@@ -154,35 +167,44 @@ def main():
     log.info("nodewatcher startup")
     instance_id = getInstanceId()
     hostname = getHostname()
-    region, asg_name, scheduler, proxy_config = getConfig(instance_id)
+    region, asg_name, scheduler, proxy_config, idle_time = getConfig(instance_id)
+
 
     s = loadSchedulerModule(scheduler)
 
+    if os.path.isfile(_NW_IDLETIME_FILE):
+        with open(_NW_IDLETIME_FILE) as f:
+            data = json.loads(f.read())
+    else:
+        data = {_NW_CURRENT_IDLETIME: 0}
+
+    atexit.register(saveIdleTime, data)
     while True:
         time.sleep(60)
-        ec2_conn = boto3.resource('ec2', region_name=region, config=proxy_config)
         asg_conn = boto3.client('autoscaling', region_name=region, config=proxy_config)
-        hour_percentile = getHourPercentile(instance_id, ec2_conn)
-        log.info('Percent of hour used: %d' % hour_percentile)
-
-        if hour_percentile < 95:
-            continue
 
-        jobs = getJobs(s, hostname)
-        if jobs == True:
+        has_jobs = hasJobs(s, hostname)
+        if has_jobs:
             log.info('Instance has active jobs.')
+            data[_NW_CURRENT_IDLETIME] = 0
         else:
             if maintainSize(asg_name, asg_conn):
                 continue
-            # avoid race condition by locking and verifying
-            lockHost(s, hostname)
-            jobs = getJobs(s, hostname)
-            if jobs == True:
-                log.info('Instance actually has active jobs.')
-                lockHost(s, hostname, unlock=True)
-                continue
             else:
-                selfTerminate(asg_name, asg_conn, instance_id)
+                data[_NW_CURRENT_IDLETIME] += 1
+                log.info('Instance %s has no job for the past %s minute(s)' % (instance_id, data[_NW_CURRENT_IDLETIME]))
+                if data[_NW_CURRENT_IDLETIME] >= idle_time:
+                    lockHost(s, hostname)
+                    has_jobs = hasJobs(s, hostname)
+                    if has_jobs:
+                        log.info('Instance has active jobs.')
+                        data[_NW_CURRENT_IDLETIME] = 0
+                    else:
+                        has_pending_jobs, error = hasPendingJobs(s)
+                        if not error and not has_pending_jobs:
+                            os.remove(_NW_IDLETIME_FILE)
+                            selfTerminate(asg_name, asg_conn, instance_id)
+                    lockHost(s, hostname, unlock=True)
```

Comment:
I think we should avoid to unlock the node in the case we're terminating it in order to avoid a time window where new jobs may get scheduled to this host.